File: server/TerraformExplorer/Services/TerraformService.cs
using TerraformExplorer.Models;
using TerraformExplorer.Models.Requests;
using TerraformExplorer.Models.Responses;
using TerraformExplorer.Settings;
using TerraformExplorer.Utils;

namespace TerraformExplorer.Services;

public class TerraformService
{
    private readonly TerraformSettings _terraformSettings;

    public TerraformService(TerraformSettings terraformSettings)
    {
        _terraformSettings = terraformSettings;
    }

    public TerraformStructure GetStructure()
    {
        return TerraformStructureLoader.Load(_terraformSettings);
    }
        
    public async Task<CommandResponse> ExecuteCommand(CommandRequest request)
    {
        return await TerraformCommandExecutor.ExecuteSingleAsync(request.Command, request.WorkingDir, _terraformSettings);
    }
    
    public async Task<ExecuteAllResponse> ExecuteAll(ExecuteAllRequest request)
    {
        return await TerraformCommandExecutor.ExecuteAllAsync(request, _terraformSettings);
    }
    
}

File: apps/web/src/services/TerraformService.ts
import HttpClient from "@/clients/HttpClient";
import type { CommandRequest } from "@/interfaces/requests/CommandRequest";
import type { ExecuteAllRequest } from "@/interfaces/requests/ExecuteAllRequest";
import type { CommandResponse } from "@/interfaces/responses/CommandResponse";
import type { ExecuteAllResponse } from "@/interfaces/responses/ExecuteAllResponse";
import type { TerraformStructure } from "@/interfaces/TerraformStructure";

export class TerraformService {
  private static httpClient: HttpClient;
  private static baseURL: string = `${import.meta.env.VITE_API_BASE_URL}/terraform`;

  private static getHttpClient(): HttpClient {
    if (!this.httpClient) {
      this.httpClient = new HttpClient(this.baseURL);
    }
    return this.httpClient;
  }

  static async getStructure(): Promise<TerraformStructure> {
    return this.getHttpClient().get<TerraformStructure>("structure");
  }

  static async executeCommand(request: CommandRequest): Promise<CommandResponse> {
    return this.getHttpClient().post<CommandResponse>("execute", request);
  }

  static async executeAll(request: ExecuteAllRequest): Promise<ExecuteAllResponse> {
    return this.getHttpClient().post<ExecuteAllResponse>("execute-all", request);
  }
}